docs/markdown/snippets/thread_model.md

```diff
@@ -0,0 +1,27 @@
+## Expose the thread model from dependency('threads')
+
+Meson now exposes the threading model through the get_variable method.
+
+```meson
+dep_threads = dependency('threads', method : 'system')
+model = dep_threads.get_variable(meson : 'model')
+if model == 'win32'
+  ...
+elif model == 'posix'
+  ...
+else
+  ...
+endif
+```
+
+This is also exposed if CMAKE is used to find threads, but due to cmake's
+implementation it is a bit different:
+
+```meson
+dep_threads = dependency('threads', method : 'cmake')
+if dep_threads.get_variable(cmake : 'CMAKE_USE_WIN32_THREADS_INIT') == '1':
+   # using windows threads
+elif dep_threads.get_variable(cmake : 'CMAKE_USE_PTHREADS_INIT') == '1':
+   # using pthreads.
+endif
+```
```

Comment:
Having two different ways of querying this information is not nice. There should only be one that works with all methods. Maybe we could do something like:

```
thread_dep = dependency('threads')
if thread_dep.threading_type() == 'posix'
  ...
endif
```

And where calling `threading_type` on a dependency that is not an actual thread dep would be a hard error?

Reply:
I really dislike adding special methods to generic types that only work on one specific class.

What if you have code like:
```meson
dep_threads = dependency('', required : false)
if <something>
    dep_threads = dependency('threads')
endif
....

if thread_dep.threading_type() == 'posix'
   # what happens here?
endif
```

Do we implement the method for all dependency objects, and return a value like `''`? That sounds like a really good way to cause hard to detect bugs. If we don't implement it though, in the case above you *have* to ensure that `dep_thread` is actually a thread dep. That takes away some of the nice polymorphic properties. But with my proposal you could do:

```meson
dep_threads = dependency('', required : false)
if <something>
    dep_threads = dependency('threads')
endif

if dep_threads.get_variable(meson : 'model', cmake : 'CMAKE_USE_PTHREADS_INIT', default_value : '0') in ['posix', '1']
  ...
if dep_threads.get_variable(meson : 'model', cmake : 'CMAKE_USE_WIN#@_ THREADS_INIT', default_value : '0') in ['win32', '1']
  ...
endif
```

Reply:
OTOH I dislike adding kwargs that are unintuitive and only used for a special case, especially if they user needs to do extra work to get the thing they really want. One approach could be to be able to specify an value that is returned if the dependency is not a thread dep. Something along the lines of:

```meson
# Returns 'something' if the dependency is not a thread dep.
t = some_dep.threading_type('something')
```

Reply:
Realistically though when are you going to use cmake for thread dependencies? meson will *always* return something using the built-in finder unless you specifically ask it to use cmake. that means for 99% of users:
```meson
thread_model = dep_threads.get_variable(meson : 'model')
if thread_model == 'posix'
   ...
elif thread_model == 'win32'
  ...
endif
```
is completely sufficient.

Just because we haven't yet found other uses for the `meson` keyword to get_variable doesn't mean we wont. We wrap up a lot of dependenices in meson with their own special finder methods, I suspect we could use it to expose a mess of stuff from boost, for example.

Reply:
The reason I don't like adding general functionality that might get used in the future is that surprisingly often it will never be used. This is exactly YAGNI. If there are at least two (preferably three) use cases that are known and added, then it is of course a different case altogether.

Reply:
I just ran into this issue. SciPy code is using `std::thread`, and on Windows with MSVC the build failed because everything assumed that _if_ the threads dependency is available, `pthread.h` should be included. 

> If there are at least two (preferably three) use cases that are known and added, then it is of course a different case altogether.

Perhaps to break the deadlock: I think this "what's the specific implementation" API would also be helpful for better BLAS and LAPACK support - those have many flavors, and we'd like a generic detection for them (xref https://github.com/mesonbuild/meson/issues/2835#issuecomment-792600589). And then we do need to know which one was found, because they for example use different Fortran ABI's.

Reply:
Good, now we have an actual use case. What is it that you actually would need from this? And would something simple like this work:

```meson
dep = dependency('threads')
if dep.get_variable('threading_type', 'dummyfallback') == ...
```

Reply:
The SciPy code uses `std::thread`, so is generic (the easy case here compared to C). So all I need is for my code in `meson.build` to emit a single define that tells the C++ whether to include `pthread.h`. It looks like:
```C++
#include <thread>
#ifdef POCKETFFT_PTHREADS
#  include <pthread.h>
#endif
```
So something like:
```meson
if dep.get_variable('threading_type') == 'pthread'`
  cpp_flags += '-DPOCKETFFT_PTHREADS'
endif
```
would be enough.

I'd be happy with any syntax here. Reading the above thread, the one thing that would be concerning is if the code I'd have to write in `meson.build` would need to know whether it was Meson or CMake having done the detection - that seems like something that should be handled internally in Meson and not leak to the user.

Here is how I had to (unsatisfactorily) work around the problem now, by simply assuming `pthreads` is not installed when building with MSVC (because I think the regular `<thread>` is always available): https://github.com/scipy/scipy/pull/16957/commits/027635e2cca032d0596c228483424c1939c8c005

In the commit I point to I could have worked a bit harder and try-compile some code that does `#include <pthread.h>`, but that seems to defeat the purpose of the generic thread dependency a bit.